packages/cli/command_parser.pony

```diff
@@ -0,0 +1,340 @@
+        if os.required then
+          return SyntaxError(os.name, "missing value for required option")
+        end
+      end
+    end
+
+    // Check for missing args and error if found.
+    while arg_pos < _spec.args.size() do
+      try
+        let ars = _spec.args(arg_pos)
+        if ars.required then
+          return SyntaxError(ars.name, "missing value for required argument")
+        end
+        args.update(ars.name, Arg(ars, ars.default))
+      end
+      arg_pos = arg_pos + 1
+    end
+
+    // A successfully parsed and populated leaf Command
+    Command(_spec, options, args)
+
+  fun box _parse_long_option(
+    token: String,
+    args: Array[String] ref): (Option | SyntaxError)
+  =>
+    """
+    --opt=foo => --opt has argument foo
+    --opt foo => --opt has argument foo, iff arg is required
+    """
+    let parts = token.split("=")
+    let name = try parts(0) else "???" end
+    let targ = try parts(1) else None end
+    match _option_with_name(name)
+    | let fs: OptionSpec => OptionParser.parse(fs, targ, args)
+    | None => SyntaxError(name, "unknown long option")
+    else
+      // TODO: Ponyc should know we've covered all match cases above
+      SyntaxError(name, "Pony shouldn't allow this")
+    end
+
+  fun box _parse_short_options(
+    token: String,
+    args: Array[String] ref): (Array[Option] | SyntaxError)
+  =>
+    """
+    if 'O' requires an argument
+      -OFoo  => -O has argument Foo
+      -O=Foo => -O has argument Foo
+      -O Foo => -O has argument Foo
+    else
+      -O=Foo => -O has argument foo
+    -abc => options a, b, c.
+    -abcFoo => options a, b, c. c has argument Foo iff its arg is required.
+    -abc=Foo => options a, b, c. c has argument Foo.
+    -abc Foo => options a, b, c. c has argument Foo iff its arg is required.
+    """
+    let parts = token.split("=")
+    let shorts = (try parts(0) else "" end).clone()
+    var targ = try parts(1) else None end
+
+    let options: Array[Option] ref = options.create()
+    while shorts.size() > 0 do
+      let c = try shorts.shift() else 0 end  // Should never error since checked
+      match _option_with_short(c)
+      | let fs: OptionSpec =>
+        if fs._requires_arg() and (shorts.size() > 0) then
+          // opt needs an arg, so consume the remainder of the shorts for targ
+          if targ is None then
+            targ = shorts.clone()
+            shorts.truncate(0)
+          else
+            return SyntaxError(short_string(c), "ambiguous args for short option")
+          end
+        end
+        let arg = if shorts.size() == 0 then targ else None end
+        match OptionParser.parse(fs, arg, args)
+        | let f: Option => options.push(f)
+        | let se: SyntaxError => return se
+        end
+      | None => SyntaxError(short_string(c), "unknown short option")
+      else
+        // TODO: Ponyc should know we've covered all match cases above
+        return SyntaxError(token, "Pony shouldn't allow this")
```

Comment:
It just got merged, BTW - so it is on master, and it will be in the next release.